torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -883,6 +907,36 @@ def _post_backward_hook(
                 # Delay using sharded gradient views until after the
                 # reduce-scatter instead of immediately after resharding
                 handle._use_sharded_grad_views()
+                if handle._has_optim_in_backward:
+                    for orig_param in handle.flat_param._params:
+                        # checking grad for None also filters out params
+                        # that don't belong to this rank
+                        if orig_param.grad is not None and hasattr(
+                            orig_param, "_in_backward_optimizers"
+                        ):
+                            # TODO (rohan-varma): For CPU offload, this unfortunately
+                            # operates on CPU, because the parameters and gradients
+                            # have already been offloaded.
+                            for optim in orig_param._in_backward_optimizers:
+                                # _p_assert(orig_param.grad is not None, "Expected orig param grad to not be None")
+                                # print(f"RV: calling step! {orig_param.grad.device} {orig_param.device}")
+                                # if dist.get_rank() == 0:
+                                # print(f"RV: calling step with param {orig_param} and grad {orig_param.grad} ")
+                                optim.step()
+                            # Set orig param gradient to None to maintain the
+                            # invariant that if flat_param.grad is None, then
+                            # each orig param grad is None.
+                            orig_param.grad = None
+                    # flat_param.grad is set to None for sharded strategies, set it here
+                    # for no_shard.
+                    # TODO (rohan-varma): see if this can be absorbed into a sharded_grad setter,
+                    # similar to sharded_grad property.
+                    handle.flat_param.grad = None
+                    handle.flat_param._saved_grad_shard = None
+                    if handle._offload_params:
+                        handle.flat_param._cpu_grad = None
```

Comment:
Relating to above comments, is this set to `None` only for the purpose of an invariant? What value do we get from doing this? In FSDP previously without optimizer overlap, I think we never set `_cpu_grad = None`?

(Feel free to only respond in one place.)